src/any_agent/frameworks/langchain.py

```diff
@@ -1,29 +1,246 @@
 
     langchain_available = True
 except ImportError:
     langchain_available = False
 
 if TYPE_CHECKING:
-    from langchain_core.language_models import LanguageModelLike
+    from langchain_core.callbacks import (
+        AsyncCallbackManagerForLLMRun,
+        CallbackManagerForLLMRun,
+    )
+    from langchain_core.language_models import LanguageModelInput, LanguageModelLike
+    from langchain_core.runnables import Runnable
+    from langchain_core.tools import BaseTool
     from langgraph.graph.state import CompiledStateGraph
+    from pydantic import BaseModel
+
+
+class ChatAnyLLM(BaseChatModel):
+    """Chat model that uses the AnyLLM API."""
+
+    model: str
+    api_key: str | None
+    api_base: str | None
+    model_kwargs: Any
+
+    def _generate(
+        self,
+        messages: list[BaseMessage],
+        stop: list[str] | None = None,
+        run_manager: CallbackManagerForLLMRun | None = None,
+        stream: bool = False,
+        **kwargs: Any,
+    ) -> ChatResult:
+        if stream:
+            stream_iter = self._stream(
+                messages, stop=stop, run_manager=run_manager, **kwargs
+            )
+            return generate_from_stream(stream_iter)
+
+        message_dicts, params = self._create_message_dicts(messages, stop)
+        params = {**params, **kwargs}
+        response = completion(messages=message_dicts, **params)  # type: ignore[arg-type]
+        if not isinstance(response, ChatCompletion):
+            error_message = f"Expected ChatCompletion, got {type(response)}"
+            raise ValueError(error_message)
+        return self._create_chat_result(response)
+
+    def _create_chat_result(self, response: ChatCompletion) -> ChatResult:
+        resp_dict = response.model_dump()
+
+        generations = []
+        token_usage = response.usage
+        for res in resp_dict["choices"]:
+            message = _convert_dict_to_message(res["message"])
+            if isinstance(message, AIMessage) and token_usage:
+                message.response_metadata = {"model_name": self.model}
+                message.usage_metadata = UsageMetadata(
+                    input_tokens=token_usage.prompt_tokens,
+                    output_tokens=token_usage.completion_tokens,
+                    total_tokens=token_usage.prompt_tokens
+                    + token_usage.completion_tokens,
+                )
+            gen = ChatGeneration(
+                message=message,
+                generation_info={"finish_reason": res.get("finish_reason")},
+            )
+            generations.append(gen)
+
+        llm_output = {
+            "token_usage": token_usage,
+            "model": self.model,
+        }
+        return ChatResult(generations=generations, llm_output=llm_output)
+
+    def _create_message_dicts(
+        self, messages: list[BaseMessage], stop: list[str] | None = None
+    ) -> tuple[list[dict[str, Any]], dict[str, Any]]:
+        params = {
```

Comment:
nit: I am a little confused why `params` is handled inside this `_create_message_dicts`, seem 2 completely different things.